x/btcstaking/keeper/msg_server.go

```diff
@@ -437,19 +313,134 @@ func (ms msgServer) CreateBTCDelegation(goCtx context.Context, req *types.MsgCre
+	}, nil
+}
+
+// CreateBTCDelegation creates a BTC delegation
+func (ms msgServer) CreateBTCDelegation(goCtx context.Context, req *types.MsgCreateBTCDelegation) (*types.MsgCreateBTCDelegationResponse, error) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), types.MetricsKeyCreateBTCDelegation)
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// 1. Parse the message into better domain format
+	parsedMsg, err := types.ParseCreateDelegationMessage(req)
+
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "%v", err)
+	}
+
+	// 2. Basic stateless checks
+	// - verify proof of possession
+	if err := parsedMsg.ParsedPop.Verify(parsedMsg.StakerAddress, parsedMsg.StakerPK.BbnPk, ms.btcNet); err != nil {
+		return nil, types.ErrInvalidProofOfPossession.Wrapf("error while validating proof of posession: %v", err)
+	}
+
+	// 3. Check if it is not duplicated staking tx
+	stakingTxHash := parsedMsg.StakingTx.Transaction.TxHash()
+	delgation := ms.getBTCDelegation(ctx, stakingTxHash)
+	if delgation != nil {
+		return nil, types.ErrReusedStakingTx.Wrapf("duplicated tx hash: %s", stakingTxHash.String())
+	}
+
+	// 4. Check finality providers to which message delegate
+	// Ensure all finality providers are known to Babylon, are not slashed,
+	// and their registered epochs are finalised
+	for _, fpBTCPK := range parsedMsg.FinalityProviderKeys.PublicKeysBbnFormat {
+		// get this finality provider
+		fp, err := ms.GetFinalityProvider(ctx, fpBTCPK)
+		if err != nil {
+			return nil, err
+		}
+		// ensure the finality provider is not slashed
+		if fp.IsSlashed() {
+			return nil, types.ErrFpAlreadySlashed
+		}
+	}
+
+	// 5. Validate parsed message against parameters
+	vp := ms.GetParamsWithVersion(ctx)
+
+	btccParams := ms.btccKeeper.GetParams(ctx)
+
+	paramsValidationResult, err := ValidateParams(parsedMsg, &vp.Params, &btccParams, ms.btcNet)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// 6. Check:
+	// - timelock of staking tx
+	// - staking tx is k-deep
+	// - staking tx inclusion proof
+	stakingTxHeader := ms.btclcKeeper.GetHeaderByHash(ctx, parsedMsg.StakingTxProofOfInclusion.HeaderHash)
+
+	if stakingTxHeader == nil {
+		return nil, fmt.Errorf("header that includes the staking tx is not found")
+	}
+
+	// no need to do more validations to the btc header as it was already
+	// validate by the btclight client module
+	btcHeader := stakingTxHeader.Header.ToBlockHeader()
+
+	proofValid := btcckpttypes.VerifyInclusionProof(
+		btcutil.NewTx(parsedMsg.StakingTx.Transaction),
+		&btcHeader.MerkleRoot,
+		parsedMsg.StakingTxProofOfInclusion.Proof,
+		parsedMsg.StakingTxProofOfInclusion.Index,
+	)
+
+	if !proofValid {
+		return nil, types.ErrInvalidStakingTx.Wrapf("not included in the Bitcoin chain")
+	}
+
+	startHeight := stakingTxHeader.Height
+
+	endHeight := stakingTxHeader.Height + uint64(parsedMsg.StakingTime)
+
+	btcTip := ms.btclcKeeper.GetTipInfo(ctx)
+	stakingTxDepth := btcTip.Height - stakingTxHeader.Height
```

Comment:
```suggestion
	stakingTxDepth := btcTip.Height - stakingTxHeader.Height + 1
```

Should we follow the convention that inclusion counts as 1 confirmation?

Reply:
I think on Babylon node we are using depth everywhere i.e for being k-deep, and for w-deep. Of course we may discuss this change as it makes sense to be compatible with phase-1, but this is not something to be done in refactor pr.

For now I have created ticket to track this - https://github.com/babylonlabs-io/babylon/issues/39